map_handler.go

```diff
@@ -0,0 +1,88 @@
+package node
+
+import (
+	"errors"
+	"log"
+	"sync"
+)
+
+// Takes care of maintaining maps and insures that we know which interfaces are reachable where.
+type MapHandler interface {
+	AddConnection(NodeAddress, MapConnection)
+	FindConnection(NodeAddress) (NodeAddress, error)
+}
+
+type taggedMap struct {
+	address NodeAddress
+	new_map ReachabilityMap
+}
+
+type mapImpl struct {
+	me         NodeAddress
+	l          *sync.Mutex
+	conns      map[NodeAddress]MapConnection
+	maps       map[NodeAddress]ReachabilityMap
+	merged_map ReachabilityMap
+}
+
+func newMapImpl(me NodeAddress) MapHandler {
+	conns := make(map[NodeAddress]MapConnection)
+	maps := make(map[NodeAddress]ReachabilityMap)
+	impl := &mapImpl{me, &sync.Mutex{}, conns, maps, NewSimpleReachabilityMap()}
+	impl.merged_map.AddEntry(me)
+	return impl
+}
+
+func (m *mapImpl) addMap(update taggedMap) {
+	m.l.Lock()
+	defer m.l.Unlock()
+	m.maps[update.address].Merge(update.new_map)
+	m.merged_map.Merge(update.new_map)
+	for addr, conn := range m.conns {
+		if addr != update.address {
+			conn.SendMap(update.new_map)
+		}
+	}
+}
+
```

Comment:
So we maintain a set of individual maps per connection and one big map containing all the reachability info?